### PR TITLE
[AIRFLOW-5162] GCS Hook Upload Method Improvement

### PR DESCRIPTION
**Pull request by steven-freed**
**Jira**

- [x] My PR addresses the following Airflow Jira issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
- https://issues.apache.org/jira/browse/AIRFLOW-5162

**Description**
- [x] Here are some details about my PR, including screenshots of any UI changes:
- Allows developers to upload objects to a GCS bucket by either a local file's file name or by a file's data (string or bytes). This may help prevent issues related to crash consistency as well as help developers have a larger variety of uploading options to GCS.

**Tests**
- [x] My PR adds the following unit tests OR does not need testing for this extremely good reason:
- The changes made to the 'upload' method still allow for old functionality, while supporting new functionality for Google GCS API methods. 

**Commits**
- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "How to write a good git commit message":

Subject is separated from body by a blank line
Subject is limited to 50 characters (not including Jira issue reference)
Subject does not end with a period
Subject uses the imperative mood ("add", not "adding")
Body wraps at 72 characters
Body explains "what" and "why", not "how"

**Documentation**
- In case of new functionality, my PR adds documentation that describes how to use it.
All the public functions and the classes in the PR contain docstrings that explain what it does
If you implement backwards incompatible changes, please leave a note in the Updating.md so we can assign it to a appropriate release

**Code Quality**
- [x] Passes flake8

### CODE REVIEW

File: airflow/contrib/hooks/gcs_hook.py [hunk 1]
```diff
@@ -175,7 +175,30 @@ def download(self, bucket_name, object_name, filename=None):
 
         return blob.download_as_string()
 
-    def upload(self, bucket_name, object_name, filename,
```
Comment (ryanyuan) [inline airflow/contrib/hooks/gcs_hook.py]: Is this going to break the operators that are calling gcs_hook.upload()?

Comment (steven-freed) [reply airflow/contrib/hooks/gcs_hook.py]: Yes it would because the new version would replace 'upload' with 'upload_from_filename' and 'upload_from_string'. That is a very good point thank you. I have modified the upload method to be more generic for both types of uploads.

**Comment by mik-laj** [comment]
Please follow pull request's guidelines: https://github.com/apache/airflow/blob/master/CONTRIBUTING.md#pull-request-guidelines
This is required for your PR to be accepted.  Without organization of work and Jira, we will not be able to manage such a large project

**Comment by steven-freed** [comment]
My PR Travis tests were canceled. I may have ran too many tests (this is my first PR). Is there any way that I can have my canceled tests run in Travis without having to push an unnecessary commit to auto run the tests?

**Comment by mik-laj** [comment]
@steven-freed I recommend configuring Travis on your own fork. 
https://github.com/apache/airflow/blob/master/CONTRIBUTING.md#testing-on-travis-ci

You do not have permission to restart jobs yourself. You need the help of the commiter.  A lot of commiters are active on Slack and can do it if you write on the #flowflow-how-to-pr channel.

The problem with the stopped task may be due to Travis overload. It is very unstable lately and we are planning to migrate to another service

I restarted this jobs.

File: airflow/contrib/hooks/gcs_hook.py [hunk 2]
```diff
@@ -175,40 +175,55 @@ def download(self, bucket_name, object_name, filename=None):
...
         :param object_name: The object name to set when uploading the local file.
         :type object_name: str
         :param filename: The local file path to the file to be uploaded.
         :type filename: str
+        :param data: The file's content as a string or bytes to be uploaded.
+        :type data: str
         :param mime_type: The MIME type to set when uploading the file.
         :type mime_type: str
-        :param gzip: Option to compress file for upload
+        :param gzip: Option to compress local file for upload
```
Comment (mik-laj) [inline airflow/contrib/hooks/gcs_hook.py]: Why didn't you add compression support for data?
```
            if gzip:
                out = StringIO.StringIO()
                with gz.GzipFile(fileobj=out, mode="w") as f:
                    f.write(data)
                data = out.getvalue()
```

Comment (steven-freed) [reply airflow/contrib/hooks/gcs_hook.py]: I actually had a use case at work where compression couldn’t be used when uploading files to BigQuery from a Storage Bucket. I didn’t think to add it but that is a great idea. I will add it asap. Thank you

**Comment by mik-laj** [comment]
Can you add tests to prevent regression?

**Comment by steven-freed** [comment]
Can you point me in the right direction of writing tests with GCP? I saw in the airflow tests directory there were tests for hooks but none for GCS. I am a bit confused on what credentials to use to mock a GCP account for tests.

**Comment by mik-laj** [comment]
Unit tests for GCS Hook is available here:
https://github.com/apache/airflow/blob/master/tests/contrib/hooks/test_gcs_hook.py

When it comes to writing tests for GCP integration, we distinguish two types:
* Unit tests -  It uses very actively mocking. You should never use authorization data in these tests. Method calls from the Google library should be mocked.
* System tests - It uses real DAG examples and uses authorization data. Unfortunately, they are not yet automatically launched on CI.

Hooks and operators are tested using unit tests, but whole integration is tested using system tests. I hope you will write unit tests.
System tests for integration with GCS have not yet been approved, but there is open PR: 
https://github.com/apache/airflow/pull/5951
My team has been actively working over the past two weeks to increase system test coverage.

**Comment by potiuk** [comment]
Hello @steven-freed -> can you please rebase and fix the one failing test? I would gladly review and merge that one.

**Comment by steven-freed** [comment]
Yes I will I apologize I just started up my course work and have been busy but I will this weekend.

**Comment by potiuk** [comment]
No worries @steven-freed  - glad that you are contributing it :)

**Comment by steven-freed** [comment]
After rebasing now I am getting errors for "Tests sqlite python 3.5" that I cannot seem to figure out. I haven't changed anything in those files, it seems like its coming from another module. I believe I rebased it correctly, maybe I am wrong...

ERROR - Failed to import plugin /opt/airflow/tests/plugins/test_plugin.py
{plugins_manager.py:148} ERROR - invalid syntax (convert.py, line 80)

**Comment by potiuk** [comment]
This is the error from Travis. So it does seem to be related to your change. What I recommend - setup Breeze (see BREEZE.rst) and try to run/fix this test 



```
======================================================================
12) FAIL: test_upload_string (tests.contrib.hooks.test_gcs_hook.TestGoogleCloudStorageHookUpload)
----------------------------------------------------------------------
   Traceback (most recent call last):
    /usr/local/lib/python3.6/unittest/mock.py line 1183 in patched
      return func(*args, **keywargs)
    tests/contrib/hooks/test_gcs_hook.py line 633 in test_upload_string
      content_type='text/plain'
    /usr/local/lib/python3.6/unittest/mock.py line 824 in assert_called_once_with
      raise AssertionError(msg)
   AssertionError: Expected 'upload_from_string' to be called once. Called 0 times.
   -------------------- >> begin captured logging << --------------------
   airflow.contrib.hooks.gcs_hook.GoogleCloudStorageHook: INFO: Data stream uploaded to test_object in test_bucket bucket
   --------------------- >> end captured logging << ---------------------
=====================================================================
```

**Comment by steven-freed** [comment]
The error I’m talking about is actually on my forked repos master. I was trying to test on my own Travis before pushing a new change to my gas-hook-update branch. I know the error above was my code that was what I wanted to fix. But now I have that new error I pointed out that occurred after I rebased my forked repos master.

**Comment by potiuk** [comment]
Can you please provide a link ? Are you rebased onto most recent apache/master ?

**Comment by steven-freed** [comment]
git remote show origin is saying all branches are tracked and my master is up to date. Heres a link to my fork and travis.
https://github.com/steven-freed/airflow
https://travis-ci.com/steven-freed/airflow/builds/126309005

I got it now I'll fix that test. Just recloned, my local repo wasn't working. Thank you for your patience with me as I am new to pull requests!

**Comment by steven-freed** [comment]
Everything looks ready to go. Please let me know if I can do anything else :)

File: tests/contrib/hooks/test_gcs_hook.py [hunk 3]
```diff
@@ -602,21 +604,103 @@ def test_upload(self, mock_service):
...
+                                        data=self.testdata_str,
+                                        gzip=True)
+
+        self.assertIsNone(response)
+
+        response = self.gcs_hook.upload(test_bucket,  # pylint:disable=assignment-from-no-return
+                                        test_object,
+                                        data=self.testdata_bytes,
+                                        gzip=True)
+
```
Comment (potiuk) [inline tests/contrib/hooks/test_gcs_hook.py]: I think we need to run asserr_called_once_with here.

File: tests/contrib/hooks/test_gcs_hook.py [hunk 4]
```diff
@@ -602,21 +604,103 @@ def test_upload(self, mock_service):
...
+    @mock.patch(GCS_STRING.format('GoogleCloudStorageHook.get_conn'))
+    def test_upload_string_gzip(self, mock_service):
+        test_bucket = 'test_bucket'
+        test_object = 'test_object'
+
+        upload_method = mock_service.return_value.get_bucket.return_value\
+            .blob.return_value.upload_from_string
+        upload_method.return_value = None
+
+        response = self.gcs_hook.upload(test_bucket,  # pylint:disable=assignment-from-no-return
```
Comment (potiuk) [inline tests/contrib/hooks/test_gcs_hook.py]: I think it's not really needed to add response and assertIsNone() (and add #pylint exceptions). It does not add any value. What you should do instead is to add type annotation for the upload_method() and add -> None there and it will be much better way to make sure it behaves as it should (via static checks rather than tests). Adding type annotations is a good thing actually

File: airflow/contrib/hooks/gcs_hook.py [hunk 5]
```diff
@@ -185,40 +186,64 @@ def download(self, bucket_name, object_name, filename=None):
         else:
             return blob.download_as_string()
 
-    def upload(self, bucket_name, object_name, filename,
-               mime_type='application/octet-stream', gzip=False):
+    def upload(self, bucket_name, object_name, filename=None,
```
Comment (potiuk) [inline airflow/contrib/hooks/gcs_hook.py]: Can you please add type annotations here? see the comment below - that would be much better than assertNone and also it's great for users of the hook.

File: tests/contrib/hooks/test_gcs_hook.py [hunk 6]
```diff
@@ -602,21 +604,103 @@ def test_upload(self, mock_service):
...
+    @mock.patch(GCS_STRING.format('GoogleCloudStorageHook.get_conn'))
+    def test_upload_exceptions(self, mock_service):
+        test_bucket = 'test_bucket'
+        test_object = 'test_object'
+
+        upload_method = mock_service.return_value.get_bucket.return_value\
+            .blob.return_value.upload_from_string
+        upload_method.return_value = None
+
+        self.assertRaises(ValueError,
```
Comment (potiuk) [inline tests/contrib/hooks/test_gcs_hook.py]: It's not really nice way of using asserRaises. It is difficult to read. I think it should look similar to:

```
with self.assertRaises(ValueError) as cm:
    self.gcs_hook.upload(test_bucket,test_object)

the_exception = cm.exception
self.assertEqual("... Message here...", the_exception.message)
```

**Review by potiuk — CHANGES_REQUESTED** [review]
There are still few comments - I think the test code might be much more readable (I proposed a few changes) and some assertions can be skipped in favour of Type Annotation.

**Comment by steven-freed** [comment]
I got everything working but seem to have a build documentation error. I can't seem to figure it out. I am up to date with the master.

File: airflow/gcp/hooks/gcs.py [hunk 7]
```diff
@@ -185,45 +180,72 @@ def download(self, bucket_name, object_name, filename=None):
...
 
-    def upload(self, bucket_name, object_name, filename,
-               mime_type='application/octet-stream', gzip=False):
+    def upload(self, bucket_name: str, object_name: str, filename: str = None,
+               data: Union[str, bytes] = None, mime_type: str = None, gzip: bool = False,
+               encoding: str = 'utf-8') -> None:
         """
-        Uploads a local file to Google Cloud Storage.
-
+        Uploads a local file or file data as string or bytes to Google Cloud Storage.
```
Comment (mik-laj) [inline airflow/gcp/hooks/gcs.py:193]: New line is required after method description. Can you add it?

File: airflow/gcp/hooks/gcs.py [hunk 8]
```diff
@@ -185,45 +180,72 @@ def download(self, bucket_name, object_name, filename=None):
...
+            blob.upload_from_string(data,
+                                    content_type=mime_type)
+            self.log.info('Data stream uploaded to %s in %s bucket', object_name, bucket_name)
+        else:
+            raise ValueError("'filename' and 'data' parameter missing. "
+                             "One is required to upload to gcs.")
 
     def exists(self, bucket_name, object_name):
         """
         Checks for the existence of a file in Google Cloud Storage.
```
Comment (mik-laj) [inline airflow/gcp/hooks/gcs.py]: This change prevents the documentation from being built properly. Can you delete it?

File: airflow/gcp/hooks/gcs.py [hunk 9]
```diff
@@ -238,7 +260,6 @@ def exists(self, bucket_name, object_name):
     def is_updated_after(self, bucket_name, object_name, ts):
         """
         Checks if an blob_name is updated in Google Cloud Storage.
-
```
Comment (mik-laj) [inline airflow/gcp/hooks/gcs.py]: This change prevents the documentation from being built properly. Can you delete it?

File: airflow/gcp/hooks/gcs.py [hunk 10]
```diff
@@ -273,7 +294,6 @@ def is_updated_after(self, bucket_name, object_name, ts):
     def delete(self, bucket_name, object_name):
         """
         Deletes an object from the bucket.
-
```
Comment (mik-laj) [inline airflow/gcp/hooks/gcs.py]: This change prevents the documentation from being built properly. Can you delete it?

File: airflow/gcp/hooks/gcs.py [hunk 11]
```diff
@@ -289,7 +309,6 @@ def delete(self, bucket_name, object_name):
     def list(self, bucket_name, versions=None, max_results=None, prefix=None, delimiter=None):
         """
         List all objects from the bucket with the give string prefix in name
-
```
Comment (mik-laj) [inline airflow/gcp/hooks/gcs.py]: This change prevents the documentation from being built properly. Can you delete it?

File: airflow/gcp/hooks/gcs.py [hunk 12]
```diff
@@ -336,13 +355,11 @@ def list(self, bucket_name, versions=None, max_results=None, prefix=None, delimi
     def get_size(self, bucket_name, object_name):
         """
         Gets the size of a file in Google Cloud Storage.
-
```
Comment (mik-laj) [inline airflow/gcp/hooks/gcs.py]: This change prevents the documentation from being built properly. Can you delete it?

File: airflow/gcp/hooks/gcs.py [hunk 13]
```diff
@@ -357,7 +374,6 @@ def get_size(self, bucket_name, object_name):
     def get_crc32c(self, bucket_name, object_name):
         """
         Gets the CRC32c checksum of an object in Google Cloud Storage.
-
```
Comment (mik-laj) [inline airflow/gcp/hooks/gcs.py]: This change prevents the documentation from being built properly. Can you delete it?

File: airflow/gcp/hooks/gcs.py [hunk 14]
```diff
@@ -405,11 +420,9 @@ def create_bucket(self,
         """
         Creates a new bucket. Google Cloud Storage uses a flat namespace, so
         you can't create a bucket with a name that is already in use.
-
         .. seealso::
             For more information, see Bucket Naming Guidelines:
             https://cloud.google.com/storage/docs/bucketnaming.html#requirements
-
```
Comment (mik-laj) [inline airflow/gcp/hooks/gcs.py]: This change prevents the documentation from being built properly. Can you delete it?

File: airflow/gcp/hooks/gcs.py [hunk 15]
```diff
@@ -405,11 +420,9 @@ def create_bucket(self,
         """
         Creates a new bucket. Google Cloud Storage uses a flat namespace, so
         you can't create a bucket with a name that is already in use.
-
```
Comment (mik-laj) [inline airflow/gcp/hooks/gcs.py]: This change prevents the documentation from being built properly. Can you delete it?

File: airflow/gcp/hooks/gcs.py [hunk 16]
```diff
@@ -418,23 +431,19 @@ def create_bucket(self,
         :type resource: dict
         :param storage_class: This defines how objects in the bucket are stored
             and determines the SLA and the cost of storage. Values include
-
```
Comment (mik-laj) [inline airflow/gcp/hooks/gcs.py]: I would also ask all other unrelated changes to the documentation to be withdrawn.
